### DISCUSSION TITLE
How to understand "the actor itself is the state"? Can I resume or restart from a panic?

### DISCUSSION THREAD

**Discussion by hanekoo** [tqwewe/kameo#86] [General]
As titled.

Currently, I have two connected actors (A --linkd--> B). The A actor can get a `PanicError` (not `None`) from the B actor. At the same time, I don't want the B actor to stop. Instead, I want to restart the B actor without creating a new one.  

Is Kameo considering supporting exposing APIs for manually managing actor lifecycles?

**Comment by tqwewe** [comment]
Actor B in your example can be restarted by returning `None` from its `Actor::on_panic` implementation. This way, it kind of supervises itself

**Reply by hanekoo** [reply]
However, I want to pass the `ActorStopReason` from Actor B up to Actor A and receive the `ActorStopReason` in `on_link_died` (instead of explicitly using `actor_ref`). Then, based on the error type, I can decide whether to restart or stop Actor B.

In essence. I want to decide whether to continue or stop the actor using a flag, and decouple this decision from the Some(ActorStopReason) and None values.

**Reply by tqwewe** [reply]
Here's some ideas:
- Send an ask request to actor A from the on_panic of actor B, with the reply being a bool indicating if it should be restarted. `actor_a.ask(ShouldRestart).await`
- you mentioned a flag, maybe you can use atomic bool for this and share it between actors.
- you could always make actor B return None from on_panic, but put it in a kind of idle state. Then actor A can either kill actor B from it's on_link_died callback, or send it a message indicating it should continue as normal.

**Reply by hanekoo** [reply]
Yes, this can partially solve the problem. 

it would be even better if it could be resolved directly using Kameo's built-in hooks. This would reduce explicit actor-to-actor communication and excessive logic code, such as implementing the `Message` trait.

**Reply by tqwewe** [reply]
I'm not sure how this can be done trivially whilst also supporting actors restarting themselves.

**Reply by hanekoo** [reply]
This might not be the right approach, as it introduces some breaking changes, and it only involves some minor adjustments：https://github.com/hanekoo/kameo/commit/699d65a2c795f659889cc41802e5a5c5cada90e3

**Reply by tqwewe** [reply]
I'm not sure I understand what this commit changes exactly. It allows the actor to return an ActorStopReason, but also `false` to not stop the actor? it seems a little confusing to return an ActorStopReason but not stop the actor

**Reply by hanekoo** [reply]
The expectation is as follows:

- Using the panic_and_stop flag to decide whether to stop the actor.
- Return ActorStopReason and pass it to actor_ref.signal_link_died(..).

The above changes did not involve link_died( the second piont) ,just merely proposed the idea of manually controlling the actor's lifecycle internally.

When I saw *the actor itself is the state*, I guessed that Kameo deliberately avoided exposing interfaces for manually controlling the lifecycle from the very beginning of its design. To be precise, it's about the state interface.

I would like to know if there are similar plans in the future.

**Reply by tqwewe** [reply]
> I would like to know if there are similar plans in the future.

I'm not sure what plans you are referring to exactly. I don't have many plans to change kameos implementation drastically at this point as I'm pretty happy with how its working currently, unless there's some strong reason to do so.

**Reply by hanekoo** [reply]
Yes, I completely agree. Kameo manages to support asynchronous features while maintaining a clean API, and adding new features should be done cautiously. However, the "supervises" feature is somewhat weak (whether for itself, between actors, or at a more granular level). I'd like to explore some possibilities, but I haven't yet thought of a particularly elegant implementation.
